File: native_runtime/runtime_text_effects.py
from __future__ import annotations

import re
import unicodedata

try:
    from .runtime_player_settings import TEXT_SPEED_PRESETS, get_safe_text_speed
except ImportError:  # pragma: no cover - exported native packages import from the same directory.
    from runtime_player_settings import TEXT_SPEED_PRESETS, get_safe_text_speed


TYPEWRITER_SENTENCE_PAUSE_MS = 260
TYPEWRITER_CLAUSE_PAUSE_MS = 140
TYPEWRITER_ELLIPSIS_PAUSE_MS = 220
TYPEWRITER_LEADING_OPENERS = "“‘\"'（([{【〔〈《「『"
TYPEWRITER_TRAILING_CLOSERS = "”’\"'）)]}】〕〉》」』"
TYPEWRITER_PERIOD_ABBREVIATIONS = {
    "mr",
    "mrs",
    "ms",
    "dr",
    "prof",
    "sr",
    "jr",
    "st",
    "vs",
    "etc",
    "e.g",
    "i.e",
    "u.s",
    "u.k",
    "no",
    "fig",
    "vol",
    "ch",
    "dept",
    "inc",
    "ltd",
    "co",
}


def is_regional_indicator_symbol(char: str) -> bool:
    if not char:
        return False
    return 0x1F1E6 <= ord(char[0]) <= 0x1F1FF


def is_typewriter_grapheme_extension(char: str) -> bool:
    if not char:
        return False
    code_point = ord(char[0])
    return (
        unicodedata.category(char[0]) in {"Mn", "Mc", "Me"}
        or 0xFE00 <= code_point <= 0xFE0F
        or 0xE0100 <= code_point <= 0xE01EF
        or 0x1F3FB <= code_point <= 0x1F3FF
    )


def get_next_typewriter_cluster_index(text: str, current_index: int) -> int:
    safe_text = str(text or "")
    safe_index = max(0, min(len(safe_text), int(current_index or 0)))
    if safe_index >= len(safe_text):
        return len(safe_text)

    next_index = safe_index + 1

    if (
        is_regional_indicator_symbol(safe_text[safe_index])
        and next_index < len(safe_text)
        and is_regional_indicator_symbol(safe_text[next_index])
    ):
        return next_index + 1

    while next_index < len(safe_text):
        current_char = safe_text[next_index]
        previous_char = safe_text[next_index - 1] if next_index > 0 else ""

        if is_typewriter_grapheme_extension(current_char):
            next_index += 1
            continue
        if previous_char == "\u200d":
            next_index += 1
            continue
        if current_char == "\u200d":
            next_index = min(len(safe_text), next_index + 2)
            continue
        break

    return next_index


def get_next_typewriter_index(text: str, current_index: int) -> int:
    safe_text = str(text or "")
    safe_index = max(0, min(len(safe_text), int(current_index or 0)))
    if safe_index >= len(safe_text):
        return len(safe_text)

    next_index = get_next_typewriter_cluster_index(safe_text, safe_index)
    current_char = safe_text[safe_index]
    next_index = include_typewriter_leading_follower(safe_text, safe_index, next_index)

    if re.match(r"[A-Za-z0-9]", current_char):
        grouped = 1
        while next_index < len(safe_text) and grouped < 3 and re.match(r"[A-Za-z0-9]", safe_text[next_index]):
            next_index = get_next_typewriter_cluster_index(safe_text, next_index)
            grouped += 1

    next_index = include_typewriter_trailing_closers(safe_text, next_index)

    while next_index < len(safe_text) and safe_text[next_index].isspace():
        next_index = get_next_typewriter_cluster_index(safe_text, next_index)
    return next_index


def include_typewriter_leading_follower(text: str, current_index: int, index: int) -> int:
    safe_text = str(text or "")
    safe_current_index = max(0, min(len(safe_text), int(current_index or 0)))
    next_index = max(0, min(len(safe_text), int(index or 0)))
    if safe_current_index >= len(safe_text) or safe_text[safe_current_index] not in TYPEWRITER_LEADING_OPENERS:
        return next_index

    while next_index < len(safe_text) and safe_text[next_index] in TYPEWRITER_LEADING_OPENERS:
        next_index = get_next_typewriter_cluster_index(safe_text, next_index)

    return get_next_typewriter_cluster_index(safe_text, next_index) if next_index < len(safe_text) else next_index


def include_typewriter_trailing_closers(text: str, index: int) -> int:
    safe_text = str(text or "")
    next_index = max(0, min(len(safe_text), int(index or 0)))
    while next_index < len(safe_text) and safe_text[next_index] in TYPEWRITER_TRAILING_CLOSERS:
        next_index += 1
    return next_index


def get_typewriter_punctuation_pause_ms(text: str, full_text: str = "") -> int:
    anchor_text = get_typewriter_pause_anchor_text(text)
    anchor_char = get_typewriter_pause_anchor_char(text)
    if re.search(r"(?:\.{3,}|…+)$", anchor_text):
        return TYPEWRITER_ELLIPSIS_PAUSE_MS
    if anchor_char == "." and is_typewriter_inline_period(anchor_text, full_text):
        return 0
    if anchor_char == "." and is_typewriter_abbreviation_period(anchor_text, full_text):
        return 0
    if anchor_char in "。！？!?.":
        return TYPEWRITER_SENTENCE_PAUSE_MS
    if anchor_char in "…—-":
        return TYPEWRITER_ELLIPSIS_PAUSE_MS
    if anchor_char in "，、；;：:,":
        return TYPEWRITER_CLAUSE_PAUSE_MS
    return 0


def get_typewriter_pause_anchor_text(text: str) -> str:
    chars = list(str(text or "").rstrip())
    while len(chars) > 1 and chars[-1] in TYPEWRITER_TRAILING_CLOSERS:
        chars.pop()
    return "".join(chars)


def get_typewriter_pause_anchor_char(text: str) -> str:
    chars = list(get_typewriter_pause_anchor_text(text))
    return chars[-1] if chars else ""


def is_typewriter_inline_period(anchor_text: str, full_text: str = "") -> bool:
    safe_anchor_text = str(anchor_text or "")
    safe_full_text = str(full_text or "")
    previous_char = safe_anchor_text[-2] if len(safe_anchor_text) >= 2 else ""
    next_char = safe_full_text[len(safe_anchor_text) : len(safe_anchor_text) + 1]
    return bool(re.match(r"[A-Za-z0-9]", previous_char) and re.match(r"[A-Za-z0-9]", next_char))


def is_typewriter_abbreviation_period(anchor_text: str, full_text: str = "") -> bool:
    safe_anchor_text = str(anchor_text or "")
    safe_full_text = str(full_text or "")
    token_match = re.search(r"([A-Za-z](?:[A-Za-z]|\.)*)\.$", safe_anchor_text)
    token = token_match.group(1).lower() if token_match else ""
    next_char = safe_full_text[len(safe_anchor_text) : len(safe_anchor_text) + 1]
    return token in TYPEWRITER_PERIOD_ABBREVIATIONS and bool(next_char and next_char.isspace())


def get_native_typewriter_step_delay_ms(speed: str, visible_text: str = "", full_text: str = "") -> int:
    safe_speed = get_safe_text_speed(speed)
    chars_per_second = TEXT_SPEED_PRESETS.get(safe_speed, TEXT_SPEED_PRESETS["normal"])
    if chars_per_second >= TEXT_SPEED_PRESETS["instant"]:
        return 0
    base_delay = max(1, int(round(1000 / max(1, chars_per_second))))
    return base_delay + get_typewriter_punctuation_pause_ms(visible_text, full_text)

File: export_player_template/runtime_text_effects.js
const typewriterGraphemeSegmenter =
  typeof Intl !== "undefined" && typeof Intl.Segmenter === "function"
    ? new Intl.Segmenter(undefined, { granularity: "grapheme" })
    : null;
const TYPEWRITER_LEADING_OPENERS = "“‘\"'（([{【〔〈《「『";
const TYPEWRITER_TRAILING_CLOSERS = "”’\"'）)]}】〕〉》」』";
const TYPEWRITER_PERIOD_ABBREVIATIONS = new Set([
  "mr",
  "mrs",
  "ms",
  "dr",
  "prof",
  "sr",
  "jr",
  "st",
  "vs",
  "etc",
  "e.g",
  "i.e",
  "u.s",
  "u.k",
  "no",
  "fig",
  "vol",
  "ch",
  "dept",
  "inc",
  "ltd",
  "co",
]);
const TYPEWRITER_STEP_DELAYS = Object.freeze({
  slow: 42,
  normal: 28,
  fast: 18,
  instant: 0,
});

export function getSafeRuntimeTextSpeed(speed) {
  return Object.hasOwn(TYPEWRITER_STEP_DELAYS, speed) ? speed : "normal";
}

export function getNextTypewriterIndex(text, currentIndex) {
  const safeText = String(text ?? "");
  const safeIndex = Math.max(0, Math.min(Number(currentIndex) || 0, safeText.length));

  if (safeIndex >= safeText.length) {
    return safeText.length;
  }

  let nextIndex = getNextCodePointIndex(safeText, safeIndex);
  const currentChar = getCodePointAtIndex(safeText, safeIndex);
  nextIndex = includeTypewriterLeadingFollower(safeText, safeIndex, nextIndex);

  if (/[A-Za-z0-9]/.test(currentChar)) {
    let grouped = 1;
    while (nextIndex < safeText.length && grouped < 3) {
      const nextChar = getCodePointAtIndex(safeText, nextIndex);
      if (!/[A-Za-z0-9]/.test(nextChar)) {
        break;
      }
      nextIndex = getNextCodePointIndex(safeText, nextIndex);
      grouped += 1;
    }
  }

  nextIndex = includeTypewriterTrailingClosers(safeText, nextIndex);

  while (nextIndex < safeText.length && /\s/.test(getCodePointAtIndex(safeText, nextIndex))) {
    nextIndex = getNextCodePointIndex(safeText, nextIndex);
  }

  return nextIndex;
}

export function includeTypewriterLeadingFollower(text, currentIndex, index) {
  const safeText = String(text ?? "");
  const currentChar = getCodePointAtIndex(safeText, currentIndex);
  let nextIndex = Math.max(0, Math.min(Number(index) || 0, safeText.length));

  if (!TYPEWRITER_LEADING_OPENERS.includes(currentChar)) {
    return nextIndex;
  }

  while (nextIndex < safeText.length && TYPEWRITER_LEADING_OPENERS.includes(getCodePointAtIndex(safeText, nextIndex))) {
    nextIndex = getNextCodePointIndex(safeText, nextIndex);
  }

  return nextIndex < safeText.length ? getNextCodePointIndex(safeText, nextIndex) : nextIndex;
}

export function includeTypewriterTrailingClosers(text, index) {
  const safeText = String(text ?? "");
  let nextIndex = Math.max(0, Math.min(Number(index) || 0, safeText.length));
  while (nextIndex < safeText.length && TYPEWRITER_TRAILING_CLOSERS.includes(getCodePointAtIndex(safeText, nextIndex))) {
    nextIndex = getNextUnicodeScalarIndex(safeText, nextIndex);
  }
  return nextIndex;
}

export function getNextCodePointIndex(text, index) {
  const safeText = String(text ?? "");
  const safeIndex = Math.max(0, Math.min(Number(index) || 0, safeText.length));
  if (safeIndex >= safeText.length) {
    return safeText.length;
  }
  if (typewriterGraphemeSegmenter) {
    for (const segment of typewriterGraphemeSegmenter.segment(safeText)) {
      const segmentEnd = segment.index + segment.segment.length;
      if (safeIndex >= segment.index && safeIndex < segmentEnd) {
        return segmentEnd;
      }
    }
  }
  return getNextTypewriterClusterIndex(safeText, safeIndex);
}

export function getNextUnicodeScalarIndex(text, index) {
  const safeText = String(text ?? "");
  const safeIndex = Math.max(0, Math.min(Number(index) || 0, safeText.length));
  if (safeIndex >= safeText.length) {
    return safeText.length;
  }
  const codePoint = safeText.codePointAt(safeIndex);
  return safeIndex + (codePoint && codePoint > 0xffff ? 2 : 1);
}

export function getCodePointAtIndex(text, index) {
  const safeText = String(text ?? "");
  const safeIndex = Math.max(0, Math.min(Number(index) || 0, safeText.length));
  return safeText.slice(safeIndex, getNextUnicodeScalarIndex(safeText, safeIndex));
}

export function getTypewriterCodePointAtIndex(text, index) {
  return String(text ?? "").codePointAt(index) ?? 0;
}

export function isRegionalIndicatorSymbol(text, index) {
  const codePoint = getTypewriterCodePointAtIndex(text, index);
  return codePoint >= 0x1f1e6 && codePoint <= 0x1f1ff;
}

export function isTypewriterGraphemeExtension(text, index) {
  const codePoint = getTypewriterCodePointAtIndex(text, index);
  return (
    (codePoint >= 0x0300 && codePoint <= 0x036f) ||
    (codePoint >= 0x1ab0 && codePoint <= 0x1aff) ||
    (codePoint >= 0x1dc0 && codePoint <= 0x1dff) ||
    (codePoint >= 0x20d0 && codePoint <= 0x20ff) ||
    (codePoint >= 0xfe00 && codePoint <= 0xfe0f) ||
    (codePoint >= 0xe0100 && codePoint <= 0xe01ef) ||
    (codePoint >= 0x1f3fb && codePoint <= 0x1f3ff)
  );
}

export function getNextTypewriterClusterIndex(text, index) {
  const safeText = String(text ?? "");
  const safeIndex = Math.max(0, Math.min(Number(index) || 0, safeText.length));
  if (safeIndex >= safeText.length) {
    return safeText.length;
  }

  let nextIndex = getNextUnicodeScalarIndex(safeText, safeIndex);

  if (isRegionalIndicatorSymbol(safeText, safeIndex) && isRegionalIndicatorSymbol(safeText, nextIndex)) {
    return getNextUnicodeScalarIndex(safeText, nextIndex);
  }

  while (nextIndex < safeText.length) {
    const previousChar = getCodePointAtIndex(safeText, Math.max(0, nextIndex - 1));
    const currentChar = getCodePointAtIndex(safeText, nextIndex);

    if (isTypewriterGraphemeExtension(safeText, nextIndex) || previousChar === "\u200d") {
      nextIndex = getNextUnicodeScalarIndex(safeText, nextIndex);
      continue;
    }

    if (currentChar === "\u200d") {
      nextIndex = getNextUnicodeScalarIndex(safeText, getNextUnicodeScalarIndex(safeText, nextIndex));
      continue;
    }

    break;
  }

  return nextIndex;
}

export function getTypewriterPunctuationPause(text, fullText = "") {
  const anchorText = getTypewriterPauseAnchorText(text);
  const anchorChar = Array.from(anchorText).at(-1) ?? "";
  if (/(\.{3,}|…+)$/.test(anchorText)) {
    return 220;
  }
  if (anchorChar === "." && isTypewriterInlinePeriod(anchorText, fullText)) {
    return 0;
  }
  if (anchorChar === "." && isTypewriterAbbreviationPeriod(anchorText, fullText)) {
    return 0;
  }
  if (/[。！？!?.]/.test(anchorChar)) {
    return 260;
  }
  if (/[………。—-]/.test(anchorChar)) {
    return 220;
  }
  if (/[，、；;：:,]/.test(anchorChar)) {
    return 140;
  }
  return 0;
}

export function getTypewriterPauseAnchorText(text) {
  const chars = Array.from(String(text ?? "").trimEnd());
  while (chars.length > 1 && TYPEWRITER_TRAILING_CLOSERS.includes(chars.at(-1))) {
    chars.pop();
  }
  return chars.join("");
}

export function getTypewriterPauseAnchorChar(text) {
  return Array.from(getTypewriterPauseAnchorText(text)).at(-1) ?? "";
}

export function isTypewriterInlinePeriod(anchorText, fullText = "") {
  const safeAnchorText = String(anchorText ?? "");
  const safeFullText = String(fullText ?? "");
  const previousChar = Array.from(safeAnchorText).at(-2) ?? "";
  const nextChar = safeFullText.slice(safeAnchorText.length, safeAnchorText.length + 1);
  return /[A-Za-z0-9]/.test(previousChar) && /[A-Za-z0-9]/.test(nextChar);
}

export function isTypewriterAbbreviationPeriod(anchorText, fullText = "") {
  const safeAnchorText = String(anchorText ?? "");
  const safeFullText = String(fullText ?? "");
  const token = (safeAnchorText.match(/([A-Za-z](?:[A-Za-z]|\.)*)\.$/)?.[1] ?? "").toLowerCase();
  const nextChar = safeFullText.slice(safeAnchorText.length, safeAnchorText.length + 1);
  return TYPEWRITER_PERIOD_ABBREVIATIONS.has(token) && /\s/.test(nextChar);
}

export function getTypewriterStepDelay(speed, visibleText = "", fullText = "") {
  const safeSpeed = getSafeRuntimeTextSpeed(speed);
  if (safeSpeed === "instant") {
    return 0;
  }
  return TYPEWRITER_STEP_DELAYS[safeSpeed] + getTypewriterPunctuationPause(visibleText, fullText);
}
